Frontend: use API to show list of polling stations on polling station select page

cikzh opened the pull request:
This PR replaces hard coded polling stations with polling stations provided by the backend. It also adds a lookup when the user enters a polling station number.

marjoleintamis commented on frontend/app/component/form/polling_station_choice/PollingStationSelector.tsx:
```diff
@@ -0,0 +1,55 @@
+import { InputField } from "@kiesraad/ui";
+import { useContext, useMemo, useState } from "react";
+
+const ERROR_MESSAGE: string = "Alleen positieve nummers toegestaan";
```
Mijn aanname is nu dat de stembureaus al zijn aangemaakt en dat je zoekt op een stembureaunummer dat al in het systeem moet staan, klopt dat?

Ik denk dat de reden dat het stembureaunummer onjuist is in dit geval minder relevant is. Als de invoer niet overeenkomt met een bestaand nummer, of de invoer nu een positief of negatief getal is, dan zou 'Dit stembureaunummer is ongeldig' of 'Dit stembureaunummer bestaat niet' altijd een goed genoeg antwoord moeten zijn. Tenzij je andere redenen kunt bedenken waarom het niet mogen invoeren van negatieve getallen relevante info is voor de gebruiker.

Dit ligt natuulijk anders wanneer een stembureau word aangemaakt: dan is het wel nodig om deze foutmelding te tonen aan de gebruiker als die een negatief getal invoert.

Aangaande 'getal' vs 'nummer': in de zin "Alleen positieve getallen toegestaan" is _getallen_ correct, omdat we het hier hebben over een fout in het getal zelf en niet over de positie van het getal in een nummering. Als het stembureaunummer niet bestaat of ongeldig is, praat je wel over een _nummer_. Zie ook [schrijfwijzer](https://www.schrijfwijzer.nl/taalvragen/verwarwoordenboek/verwarwoord/54/bedrag-cijfer-getal-nummer).

praseodym replied:
Volgens mij kunnen we hier ook het design volgen, zie mijn eerdere comment: https://github.com/kiesraad/abacus/pull/146#issuecomment-2233622270

cikzh replied:
Als ik het design letterlijk zou volgen zou er geen validatie meer nodig zijn en zou er staan `Geen stembureau gevonden met nummer asdf`. Wat mij betreft is dat een prima optie is.

jorisleker replied:
yes, sluit me aan bij de laatste comments van @praseodym en @cikzh : geen aparte validatie nodig op het al dan niet zijn van een nummer. Zou wel een maxlength op het veld zetten, maar dat zit er misschien al op.

jschuurk-kr replied:
Als we "Geen stembureau gevonden met nummer asdf" tonen en de gebruiker klikt op "Beginnen", wat moet er dan gebeuren?

cikzh replied:
Op dit moment zie je dan een validatie tooltip van het `<InputField>` element. Het `Please match the requested format`, bericht.

cikzh commented:
@praseodym 
>     * The search design doesn't match the [design in Figma](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=137-5073&t=HbflVzrg2pppVyBy-4).
Added the styling
>     * If I enter a number that is not found, I can still continue to the data entry page.
This will be fixed in #108

cikzh commented on frontend/app/component/form/polling_station_choice/PollingStationSelector.tsx:
```diff
@@ -0,0 +1,49 @@
+        fieldWidth="narrow"
+        margin={false}
+        pattern="\d+"
+        title="Alleen positieve nummers toegestaan"
```
Ended up not using the `<InputField />` error, but adding it to the custom feedback box right of the input field, which matches the design

jorisleker commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.tsx:
```diff
@@ -55,115 +52,39 @@ export function PollingStationChoiceForm() {
+                    </td>
+                    <td width="5rem">
+                      <div className="link">
+                        <IconChevronRight />
```
yes, please follow the width defined in the design

jorisleker commented:
One more thing I noticed: when I open te list of stembureaus, and then scroll down to the bottom and select one, I end up on https://c6a6a9df.kiesraad-abacus.pages.dev/1/input/21 with the page scrolled down (header is outside of the viewport)
<img width="944" alt="Scherm­afbeelding 2024-07-18 om 15 52 29" src="https://github.com/user-attachments/assets/9e98fb99-f5b0-4d01-8050-a5984f66aa4a">

jschuurk-kr commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.test.tsx:
```diff
@@ -22,13 +23,28 @@ describe("Test PollingStationChoiceForm", () => {
 
-    await user.click(submitButton);
+    // Check if the station number and name exist and are visible
+    expect(screen.getByText("20")).toBeVisible();
+    expect(screen.getByText('Stembureau "Op Rolletjes"')).toBeVisible();
+    expect(screen.getByText("21")).toBeVisible();
+    expect(screen.getByText("Testplek")).toBeVisible();
```
```suggestion
    const pollingStationList = screen.getByTestId("polling_station_list")
    expect(within(pollingStationList).getByText("20")).toBeVisible();
    expect(within(pollingStationList).getByText('Stembureau "Op Rolletjes"')).toBeVisible();
    expect(within(pollingStationList).getByText("21")).toBeVisible();
    expect(within(pollingStationList).getByText("Testplek")).toBeVisible();
```

jschuurk-kr replied:
Suggestion above does not include the required import of `within`.

cikzh replied:
Nice, thanks

jschuurk-kr commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.test.tsx:
I'm missing three tests for the feedback next to the input field:
- existing polling station: "Stembureau "Op Rolletjes""
- non-existing polling station: "Geen stembureau gevonden met nummer 22"
- not a number: "Geen stembureau gevonden met nummer asd"

jschuurk-kr commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.test.tsx:
```diff
@@ -1,7 +1,8 @@
+import { PollingStationProvider } from "@kiesraad/api";
 
 describe("Test PollingStationChoiceForm", () => {
   test("Form field entry and buttons", async () => {
```
Something is going on with the `expect` on `toBeInvalid()` in this test.

If I execute these steps in a browser myself, `aria-invalid="false"` never changes to `aria-invalid="true"`.

If I change the `expect` to `expect(pollingStation).not.toBeInvalid();`, the test fails with the following output:
```
Error: expect(element).not.toBeInvalid()

Received element is currently invalid:
  <input
  aria-invalid="false"
  id="pollingStation"
  maxlength="6"
  name="number"
  pattern="\d+"
  type="text"
  value="abc"
/>
```

So the element is invalid, even though `aria-invalid="false"`?

cikzh replied:
Now that the pattern is removed, I think this is expected behavior. If not, we should open an issue

Lionqueen94 commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.tsx:
```diff
@@ -12,8 +15,11 @@ interface PollingStationChoiceFormElement extends HTMLFormElement {
+  const { pollingStations, pollingStationsLoading } = useContext(PollingStationsContext);
+
+  const handleRowClick = (pollingStationNumber: number) => () => {
+    navigate(`./${pollingStationNumber}`);
```
```suggestion
    navigate(`./${pollingStationNumber}/recounted`);
```

Lionqueen94 commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.tsx:
```diff
@@ -12,8 +15,11 @@ interface PollingStationChoiceFormElement extends HTMLFormElement {
+    navigate(`./${pollingStationNumber}`);
   };
   function handleSubmit(event: React.FormEvent<PollingStationChoiceFormElement>) {
     event.preventDefault();
```
Line below has to be changed to ```navigate(`./${pollingStationNumber}/recounted`);```

jschuurk-kr commented:
If I don't fill in a 'stembureaunummer' en klik "Beginnen", I don't get an error message and proceed to the "recounted" page of polling location 30.

Lionqueen94 commented on frontend/app/component/form/polling_station_choice/PollingStationSelector.tsx:
```diff
@@ -0,0 +1,65 @@
+        label="Voer het nummer in:"
+        fieldWidth="narrow"
+        margin={false}
+        pattern="\d+"
```
This line needs to be removed, otherwise it will still validate the field to only digits, but without a proper message because `title` was removed.

jschuurk-kr commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.test.tsx:
```diff
@@ -22,13 +23,29 @@ describe("Test PollingStationChoiceForm", () => {
+    await user.click(submitButton);
+  });
+
+  test("Polling station list", async () => {
```
I'd prefer an `overrideOnce()` at the start of this test for the api response instead of relying on what's in `frontend/lib/api-mocks/PollingStationMockData.ts`.

In this case the mock plainly returns the data, so it's unlikely to lead to isses. For the `data_entries` ednpoint it does not. So I'd rather have the rule "always do an `overrideOnce()`", which leaves no doubt what data is used by test, than the rule "it depends on how complicated someone judged the mock implementation to be".

cikzh replied:
I opened https://github.com/kiesraad/abacus/issues/163 to address this

Lionqueen94 commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.tsx:
```diff
@@ -1,6 +1,9 @@
 import { IconChevronRight } from "@kiesraad/icon";
+import { PollingStation, PollingStationsContext } from "@kiesraad/api";
+import { PollingStationSelector } from "./PollingStationSelector";
+import { useContext } from "react";
```
```suggestion
import { useContext, FormEvent } from "react";
```

Lionqueen94 commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.tsx:
```diff
@@ -12,8 +15,11 @@ interface PollingStationChoiceFormElement extends HTMLFormElement {
+  const handleRowClick = (pollingStationNumber: number) => () => {
+    navigate(`./${pollingStationNumber}/recounted`);
   };
   function handleSubmit(event: React.FormEvent<PollingStationChoiceFormElement>) {
```
```suggestion
  function handleSubmit(event: FormEvent<PollingStationChoiceFormElement>) {
```

praseodym commented:
> > * If I enter a number that is not found, I can still continue to the data entry page.
> 
> This will be fixed in #108


> If I don't fill in a 'stembureaunummer' en klik "Beginnen", I don't get an error message and proceed to the "recounted" page of polling location 30.
> 
> (edit) Might be not in scope yet until #108. I updated that issue with different cases to consider for the input.

I'd say #108 covers just the data entry pages. This would e.g. include that going to an invalid polling station id URL should throw an error. The polling station select page is not in scope there.

#107 should cover the polling station select page, including all of its behaviour. That means that typing in "1234", seeing "Geen stembureau gevonden met nummer 1234" and then hitting Enter shouldn't continue to another page. If we think including this means that this PR will get too big, we can create a second PR as part of #107.

jschuurk-kr commented on frontend/app/component/form/polling_station_choice/PollingStationChoiceForm.test.tsx:
Came up with two more tests:
- request for existing election id but election has no polling stations
- request for non-existing election id

Behavior is the same for both cases: API returns response with empty list of polling stations. So can be covered with a single test and a comment.

Behavior of the UI on such response is showing an empty list of polling stations and no polling stations can be found. I think we need some kind of error message in the UI to inform the user.

jschuurk-kr replied:
Request for non-existing election id should result in 404 (discussed in Mattermost).

praseodym replied:
#167 was created for that.

jschuurk-kr replied:
The 404 response of a non-existing election id results in the same behavior as when the API returns an empty list, because of how the `PollingStationProvider` works. So I would like to have a test that verifies/documents this behavior.

However, the tests in `PollingStationChoiceForm.test.tsx` rely on either the default values from the mock or on setting `pollingStationsLoading` and `pollingStations` in `PollingStationsContext` as part of the `render()`.

So I think we've hit a problem I was anticipating: mixing up form tests and page tests.

So I propose a double solution:
1. Add a test with an `overrideOnce()` for the 404 response behavior.
2. Create a separate issue for refactoring the tests to split the form and page tests. This probably will take some figuring out, e.g. should form tests never use mock data and should page tests always use mock data? And it might also require some refactoring of the application code.

jschuurk-kr replied:
related: #163

jschuurk-kr replied:
Approving this PR so it can be merged. :confetti_ball: Adding the 404 test can be part of #163.

jschuurk-kr commented:
@jorisleker Current implementation shows matches immediately. So if you have polling stations with numbers `1, `12`, `123`, and you want station `123`, and you type slowly enough, then you'll first see the match on polling station 1, then on station 12, then on station 123.

Am thinking that's ok for now and that user testing will tell us if it might be confusing to some users.

jorisleker commented:
> @jorisleker Current implementation shows matches immediately. So if you have polling stations with numbers `1, `12`, `123`, and you want station `123`, and you type slowly enough, then you'll first see the match on polling station 1, then on station 12, then on station 123.
> 
> Am thinking that's ok for now and that user testing will tell us if it might be confusing to some users.

Yeah, this is ok for now. Although we could:
- On a keystroke, immediately show the 'aan het zoeken' state ([design](https://www.figma.com/design/zZlFr8tYiRyp4I26sh6eqp/Kiesraad---Abacus-optelsoftware?node-id=137-5086&t=rRDhbW7cI8fvV83K-4))
- Show the found/not found feedback with a little delay (0.2 seconds will still feel almost immediate, but make the user experience a bit nicer)
- Cancel a previous lookup on a new keystroke. This will prevent the ux to show the results for query `1` if the user enters a `2` right after that. 

Found some design issues as well, see screenhots:
- the icon isn't properly vertical aligned to the text (chrome on macos)
- there is to much padding/margin to the left of the icon
- in the success state, there is also to much padding to the left of the text
- the 'eerste/tweede invoer' badge after the name of the polling station is missing

![Scherm­afbeelding 2024-07-22 om 08 49 52](https://github.com/user-attachments/assets/a466a7ca-d228-4983-bef1-28f797e23d9a)
![Scherm­afbeelding 2024-07-22 om 08 50 29](https://github.com/user-attachments/assets/b749bfa9-5b6a-4faf-bbd3-7a64c2fb15b6)

praseodym commented on frontend/lib/util/hook/useDebouncedCallback.ts:
```diff
@@ -0,0 +1,22 @@
+      window.clearTimeout(debounceTimerRef.current);
+      debounceTimerRef.current = window.setTimeout(() => callback(...args), delay);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [delay],
```
Can we somehow not disable this lint?

jschuurk-kr replied:
You mean globally?

That's not recommended: https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528

praseodym replied:
No, I want to keep the lint active instead of disabling it.

cikzh replied:
It complains about `callback` not being a dependency. If we add `callback` as a dependency, it stops working. I think it was because the callback then gets recreated every time, causing the `debounceTimerRef` to reset (not working right now, so I can't check to be sure).

praseodym replied:
Looking at @jschuurk-kr's link I wouldn't be too sure that, even though skipping `callback` in the deps works, it is also safe. Let's look into how to improve this!

Lionqueen94 replied:
This is something we ran into all the time with react, you cannot add all dependencies, because if there is one that changes, but the effect should not run every time, you cannot add the changing dep to dependencies because it will cause a never-ending loop.

praseodym replied:
Wrapping the callback in a ref (https://github.com/kiesraad/abacus/pull/146/files#r1688191696) seems to work, not sure if this has any other implications though.

praseodym commented on frontend/lib/util/hook/useDebouncedCallback.ts:
```diff
@@ -0,0 +1,22 @@
+  callback: T,
+  delay: number,
+) {
+  const debounceTimerRef = useRef(0);
+  useEffect(() => {
+    return () => {
+      window.clearTimeout(debounceTimerRef.current);
+    };
+  }, []);
+
+  return useCallback(
+    (...args: Parameters<T>) => {
+      window.clearTimeout(debounceTimerRef.current);
+      debounceTimerRef.current = window.setTimeout(() => callback(...args), delay);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [delay],
+  );
```
```suggestion
  const callbackRef = useRef(callback);
  const debounceTimerRef = useRef(0);

  useEffect(() => {
    return () => {
      window.clearTimeout(debounceTimerRef.current);
    };
  }, []);

  return useCallback(
    (...args: Parameters<T>) => {
      window.clearTimeout(debounceTimerRef.current);
      debounceTimerRef.current = window.setTimeout(() => callbackRef.current(...args), delay);
    },
    [delay],
  );```

cikzh replied:
This actually might work, I'll test it out. Thanks!

praseodym commented:
The list of commits somehow has a bunch of duplicates in there?

cikzh commented:
> The list of commits somehow has a bunch of duplicates in there?

Good catch, I did something rebasing `main`, I'll fix it

cikzh commented:
- the 'eerste/tweede invoer' badge after the name of the polling station is missing

@jorisleker I've fixed the design issues and added a debounce on the user input. I purposely left the badge out, as @jschuurk-kr mentioned earlier in a review that we should leave out design placeholders, and the badge logic is not implemented yet.

jschuurk-kr commented:
> * the 'eerste/tweede invoer' badge after the name of the polling station is missing
> 
> 
> @jorisleker I've fixed the design issues and added a debounce on the user input. I purposely left the badge out, as @jschuurk-kr mentioned earlier in a review that we should leave out design placeholders, and the badge logic is not implemented yet.

My reasoning about the placeholders: I was starting to find it hard to look at the page, check what's in scope of the PR, identify which parts of the page had been implemented and which parts are placeholders, and then decide what to review/test for the PR. That left me with less mental bandwidth for testing/reviewing than I liked.

So for pages we're actively working on, I'm in favor of either removing placeholders or having placeholders that are clearly placeholders. In this case that could be the badge saying "invoerstatus" instead of "1ste invoer".

praseodym commented:
The PR description still says:

> The PR is still in draft, because it lacks some test coverage, but it is otherwise ready for review.

I think that's outdated and can be removed?

cikzh commented:
> > * the 'eerste/tweede invoer' badge after the name of the polling station is missing
> > 
> > @jorisleker I've fixed the design issues and added a debounce on the user input. I purposely left the badge out, as @jschuurk-kr mentioned earlier in a review that we should leave out design placeholders, and the badge logic is not implemented yet.
> 
> My reasoning about the placeholders: I was starting to find it hard to look at the page, check what's in scope of the PR, identify which parts of the page had been implemented and which parts are placeholders, and then decide what to review/test for the PR. That left me with less mental bandwidth for testing/reviewing than I liked.
> 
> So for pages we're actively working on, I'm in favor of either removing placeholders or having placeholders that are clearly placeholders. In this case that could be the badge saying "invoerstatus" instead of "1ste invoer".

In this case the Badge is typed in an object, so I'd have to add a placeholder field, or create a seperate placeholder badge, so I'll just leave it out for now. The Badges are still present in the code as commented TODO's, so they'll be trivial to find and implement, once we get to that.

praseodym commented:
Even though the UI says "SHIFT + Enter", a regular Enter keypress also submits (navigates to the data entry form). The data entry form also doesn't handle this, maybe @lkleuver has already resolved it there as part of #133?

cikzh commented:
I think all feedback is either implemented or delegated to other issues. 

> Even though the UI says "SHIFT + Enter", a regular Enter keypress also submits (navigates to the data entry form). The data entry form also doesn't handle this, maybe @lkleuver has already resolved it there as part of #133?

This is now fixed

cikzh commented:
I think all feedback is either implemented or delegated to other issues
